Foundation/JSONSerialization.swift

```diff
@@ -285,219 +454,176 @@ internal extension JSONSerialization {
+    init(pretty: Bool = false, sortedKeys: Bool = false) {
         self.pretty = pretty
         self.sortedKeys = sortedKeys
-        self.writer = writer
     }
     
-    mutating func serializeJSON(_ object: Any?) throws {
+    mutating func visit(_ any: Any?) throws {
 
-        var toSerialize = object
+        var toSerialize = any
 
         if let number = toSerialize as? _NSNumberCastingWithoutBridging {
             toSerialize = number._swiftValueOfOptimalType
         }
         
         guard let obj = toSerialize else {
-            try serializeNull()
+            try NSNull().visit(with: &self)
             return
         }
         
-        // For better performance, the most expensive conditions to evaluate should be last.
-        switch (obj) {
-        case let str as String:
-            try serializeString(str)
-        case let boolValue as Bool:
-            writer(boolValue.description)
-        case let num as Int:
-            writer(num.description)
-        case let num as Int8:
-            writer(num.description)
-        case let num as Int16:
-            writer(num.description)
-        case let num as Int32:
-            writer(num.description)
-        case let num as Int64:
-            writer(num.description)
-        case let num as UInt:
-            writer(num.description)
-        case let num as UInt8:
-            writer(num.description)
-        case let num as UInt16:
-            writer(num.description)
-        case let num as UInt32:
-            writer(num.description)
-        case let num as UInt64:
-            writer(num.description)
-        case let array as Array<Any?>:
-            try serializeArray(array)
-        case let dict as Dictionary<AnyHashable, Any?>:
-            try serializeDictionary(dict)
-        case let num as Float:
-            try serializeFloat(num)
-        case let num as Double:
-            try serializeFloat(num)
-        case let num as Decimal:
-            writer(num.description)
-        case let num as NSDecimalNumber:
-            writer(num.description)
-        case is NSNull:
-            try serializeNull()
-        case _ where __SwiftValue.store(obj) is NSNumber:
-            let num = __SwiftValue.store(obj) as! NSNumber
-            writer(num.description)
-        default:
-            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid object cannot be serialized"])
+        guard let jsonObj = obj as? _JSONValue else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid value (\(obj)) in JSON write"])
         }
+        
+        try jsonObj.visit(with: &self)
+    }
+    
+    mutating func visit<I: FixedWidthInteger> (_ int: I) {
+        result += int.description //TODO
```

Comment:
Relying on .description just doesn't seem right to me… I think we have some nice dedicated stuff for this now?

Reply:
Yeah I think we want to move away from description in general.

Reply:
Well in Obj-C I believe but not sure in Swift's case...